File: liva-flutter-app/android/liva-animation/src/main/kotlin/com/liva/animation/rendering/AnimationEngine.kt
package com.liva.animation.rendering

import android.graphics.Bitmap
import android.graphics.PointF
import android.util.Log
import com.liva.animation.models.DecodedFrame
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

private const val TAG = "AnimationEngine"

/**
 * Animation playback mode.
 */
enum class AnimationMode(val frameRate: Double) {
    IDLE(10.0),      // 10 fps for idle animations
    TALKING(30.0),   // 30 fps for lip sync
    TRANSITION(30.0); // Transition between animations

    val frameIntervalMs: Long
        get() = (1000.0 / frameRate).toLong()
}

/**
 * Frame ready for rendering.
 */
data class RenderFrame(
    val baseImage: Bitmap?,
    val overlayImage: Bitmap?,
    val overlayPosition: PointF,
    val timestamp: Long
)

/**
 * Animation chunk for queued playback.
 */
data class QueuedAnimationChunk(
    val chunkIndex: Int,
    val frames: List<DecodedFrame>,
    val overlayPosition: PointF,
    val animationName: String,
    var isReady: Boolean = false
)

/**
 * Manages animation frame timing and queue.
 */
internal class AnimationEngine {

    // MARK: - Properties

    var mode: AnimationMode = AnimationMode.IDLE
        private set

    var currentAnimationName: String = ""
        private set

    private val frameQueue = mutableListOf<DecodedFrame>()
    private val queueLock = ReentrantLock()

    private val chunkQueue = mutableListOf<QueuedAnimationChunk>()
    private val chunkLock = ReentrantLock()

    private var currentFrameIndex = 0
    private var lastFrameTime = 0L
    private var isPlaying = false

    private var baseFrame: Bitmap? = null
    private var baseFrameManager: BaseFrameManager? = null
    private var overlayPosition = PointF(0f, 0f)

    private val bufferThreshold = 10

    // MARK: - Callbacks

    var onModeChange: ((AnimationMode) -> Unit)? = null
    var onChunkComplete: ((Int) -> Unit)? = null
    var onAnimationComplete: (() -> Unit)? = null

    // MARK: - Configuration

    var minimumBufferFrames: Int = 10
    var loopIdleAnimations: Boolean = true

    // MARK: - Frame Queue Management

    /**
     * Enqueue decoded frames for playback.
     */
    fun enqueueFrames(frames: List<DecodedFrame>, chunkIndex: Int) {
        queueLock.withLock {
            frameQueue.addAll(frames)
            frameQueue.sortBy { it.sequenceIndex }

            if (frameQueue.size >= bufferThreshold && !isPlaying) {
                isPlaying = true
                setMode(AnimationMode.TALKING)
            }
        }
    }

    /**
     * Add a chunk to the queue.
     */
    fun enqueueChunk(chunk: QueuedAnimationChunk) {
        chunkLock.withLock {
            chunkQueue.add(chunk)
            chunkQueue.sortBy { it.chunkIndex }
        }
    }

    /**
     * Mark a chunk as ready.
     */
    fun markChunkReady(chunkIndex: Int) {
        chunkLock.withLock {
            chunkQueue.find { it.chunkIndex == chunkIndex }?.isReady = true
        }
    }

    /**
     * Set the overlay position.
     */
    fun setOverlayPosition(position: PointF) {
        overlayPosition = position
    }

    /**
     * Set the base frame.
     */
    fun setBaseFrame(frame: Bitmap?) {
        baseFrame = frame
    }

    /**
     * Set base frame manager for idle animations.
     */
    fun setBaseFrameManager(manager: BaseFrameManager?) {
        baseFrameManager = manager
    }

    /**
     * Set animation mode.
     */
    fun setMode(newMode: AnimationMode) {
        if (mode != newMode) {
            mode = newMode
            onModeChange?.invoke(newMode)
        }
    }

    /**
     * Clear all queued frames.
     */
    fun clearQueue() {
        queueLock.withLock {
            frameQueue.forEach { it.image.recycle() }
            frameQueue.clear()
            currentFrameIndex = 0
        }

        chunkLock.withLock {
            chunkQueue.clear()
        }

        isPlaying = false
    }

    /**
     * Transition to idle.
     */
    fun transitionToIdle() {
        setMode(AnimationMode.IDLE)
    }

    // MARK: - Frame Retrieval

    /**
     * Get the next frame for rendering.
     */
    fun getNextFrame(): RenderFrame? {
        val currentTime = System.currentTimeMillis()

        if (currentTime - lastFrameTime < mode.frameIntervalMs) {
            return getCurrentRenderFrame(currentTime)
        }

        lastFrameTime = currentTime

        // Handle idle mode with base frame manager
        if (mode == AnimationMode.IDLE) {
            val idleFrame = getIdleFrame(currentTime)
            if (idleFrame.baseImage == null) {
                Log.w(TAG, "getNextFrame IDLE returning NULL baseImage!")
            }
            return idleFrame
        }

        // Get current base frame from manager
        val currentBaseFrame = baseFrameManager?.getCurrentIdleFrame() ?: baseFrame
        if (currentBaseFrame == null) {
            Log.w(TAG, "getNextFrame TALKING: currentBaseFrame is NULL (manager=${baseFrameManager != null}, baseFrame=${baseFrame != null})")
        }

        var overlayImage: Bitmap? = null

        queueLock.withLock {
            if (frameQueue.isNotEmpty()) {
                if (currentFrameIndex < frameQueue.size) {
                    overlayImage = frameQueue[currentFrameIndex].image
                    currentFrameIndex++
                }

                if (currentFrameIndex >= frameQueue.size) {
                    if (mode == AnimationMode.TALKING) {
                        if (hasMoreChunks()) {
                            loadNextChunk()
                        } else {
                            currentFrameIndex = frameQueue.size - 1
                            overlayImage = frameQueue.lastOrNull()?.image
                        }
                    }
                }
            }
        }

        return RenderFrame(
            baseImage = currentBaseFrame,
            overlayImage = overlayImage,
            overlayPosition = overlayPosition,
            timestamp = currentTime
        )
    }

    /**
     * Get frame for idle animation from base frame manager.
     */
    private fun getIdleFrame(timestamp: Long): RenderFrame {
        val manager = baseFrameManager
        if (manager == null) {
            // Fallback to static base frame
            Log.w(TAG, "getIdleFrame: No baseFrameManager, using static baseFrame=${baseFrame != null}")
            return RenderFrame(
                baseImage = baseFrame,
                overlayImage = null,
                overlayPosition = PointF(0f, 0f),
                timestamp = timestamp
            )
        }

        // Advance and get next idle frame (10fps handled by mode.frameIntervalMs)
        val idleFrame = manager.advanceFrame()

        // Use the new frame, or fall back to current frame, or fall back to static baseFrame
        // This ensures we NEVER return null for the base image
        val frameToUse = idleFrame
            ?: manager.getCurrentIdleFrame()
            ?: baseFrame

        if (frameToUse == null) {
            Log.w(TAG, "getIdleFrame: ALL FALLBACKS FAILED! advanceFrame=${idleFrame != null}, getCurrentIdleFrame=${manager.getCurrentIdleFrame() != null}, baseFrame=${baseFrame != null}")
        }

        return RenderFrame(
            baseImage = frameToUse,
            overlayImage = null,
            overlayPosition = PointF(0f, 0f),
            timestamp = timestamp
        )
    }

    /**
     * Get current render frame without advancing (between frame intervals).
     * BUG FIX: Now properly uses baseFrameManager instead of just baseFrame.
     */
    private fun getCurrentRenderFrame(timestamp: Long): RenderFrame {
        // Get base frame from manager (like getIdleFrame does) - FIXED BUG!
        val manager = baseFrameManager
        val currentBaseFrame = manager?.getCurrentIdleFrame() ?: baseFrame

        if (currentBaseFrame == null) {
            Log.w(TAG, "getCurrentRenderFrame: NULL baseFrame! manager=${manager != null}, baseFrame=${baseFrame != null}")
        }

        queueLock.withLock {
            val overlayImage = when {
                frameQueue.isEmpty() -> null
                currentFrameIndex < frameQueue.size -> frameQueue[currentFrameIndex].image
                frameQueue.isNotEmpty() -> frameQueue.last().image
                else -> null
            }

            return RenderFrame(
                baseImage = currentBaseFrame,
                overlayImage = overlayImage,
                overlayPosition = overlayPosition,
                timestamp = timestamp
            )
        }
    }

    // MARK: - Chunk Management

    private fun hasMoreChunks(): Boolean {
        chunkLock.withLock {
            return chunkQueue.isNotEmpty() && chunkQueue.first().isReady
        }
    }

    private fun loadNextChunk() {
        val nextChunk: QueuedAnimationChunk

        chunkLock.withLock {
            if (chunkQueue.isEmpty() || !chunkQueue.first().isReady) return
            nextChunk = chunkQueue.removeAt(0)
        }

        queueLock.withLock {
            val overlapFrames = minOf(5, frameQueue.size)
            if (overlapFrames > 0) {
                val keep = frameQueue.takeLast(overlapFrames)
                frameQueue.clear()
                frameQueue.addAll(keep)
                currentFrameIndex = 0
            } else {
                frameQueue.clear()
                currentFrameIndex = 0
            }

            frameQueue.addAll(nextChunk.frames)
            overlayPosition = nextChunk.overlayPosition
            currentAnimationName = nextChunk.animationName
        }

        onChunkComplete?.invoke(nextChunk.chunkIndex)
    }

    // MARK: - Buffer Status

    val hasBufferedFrames: Boolean
        get() = queueLock.withLock { frameQueue.size >= minimumBufferFrames }

    val queuedFrameCount: Int
        get() = queueLock.withLock { frameQueue.size }

    val pendingChunkCount: Int
        get() = chunkLock.withLock { chunkQueue.size }

    val playbackProgress: Float
        get() = queueLock.withLock {
            if (frameQueue.isEmpty()) 0f
            else currentFrameIndex.toFloat() / frameQueue.size
        }

    // MARK: - Playback Control

    fun play() {
        isPlaying = true
    }

    fun pause() {
        isPlaying = false
    }

    fun reset() {
        queueLock.withLock {
            currentFrameIndex = 0
        }
        lastFrameTime = 0
    }

    val isCurrentlyPlaying: Boolean
        get() = isPlaying
}

File: liva-sdk-ios/LIVAAnimation/Sources/Rendering/AnimationEngine.swift
//
//  AnimationEngine.swift
//  LIVAAnimation
//
//  Frame timing, queue management, and playback synchronization.
//

import UIKit

/// Animation playback mode
enum LegacyAnimationMode {
    case idle       // 10 fps for idle animations
    case talking    // 30 fps for lip sync
    case transition // Transition between animations

    var frameRate: Double {
        switch self {
        case .idle: return 10.0
        case .talking: return 30.0
        case .transition: return 30.0
        }
    }

    var frameInterval: TimeInterval {
        return 1.0 / frameRate
    }
}

/// Frame ready for rendering
struct RenderFrame {
    let baseImage: UIImage?
    let overlayImage: UIImage?
    let overlayPosition: CGPoint
    let timestamp: CFTimeInterval
}

/// Animation chunk for queued playback
struct AnimationChunk {
    let chunkIndex: Int
    let frames: [DecodedFrame]
    let overlayPosition: CGPoint
    let animationName: String
    var isReady: Bool
}

/// Manages animation frame timing and queue
final class AnimationEngine {

    // MARK: - Properties

    /// Current animation mode
    private(set) var mode: LegacyAnimationMode = .idle {
        didSet {
            if oldValue != mode {
                onModeChange?(mode)
            }
        }
    }

    /// Current animation name
    private(set) var currentAnimationName: String = ""

    /// Frame queue for current chunk
    private var frameQueue: [DecodedFrame] = []
    private let queueLock = NSLock()

    /// Chunk queue for buffering
    private var chunkQueue: [AnimationChunk] = []
    private let chunkLock = NSLock()

    /// Current playback state
    private var currentFrameIndex = 0
    private var lastFrameTime: CFTimeInterval = 0
    private var isPlaying = false

    /// Base frame (full avatar when not animating mouth)
    private var baseFrame: UIImage?

    /// Base frame manager for idle/transition animations
    private weak var baseFrameManager: BaseFrameManager?

    /// Current overlay position
    private var overlayPosition: CGPoint = .zero

    /// Idle animation timer for 10fps playback
    private var idleFrameTimer: CFTimeInterval = 0

    /// Buffer threshold before starting playback
    private let bufferThreshold = 10

    // MARK: - Callbacks

    var onModeChange: ((LegacyAnimationMode) -> Void)?
    var onChunkComplete: ((Int) -> Void)?
    var onAnimationComplete: (() -> Void)?

    // MARK: - Configuration

    /// Minimum frames to buffer before playback
    var minimumBufferFrames: Int = 10

    /// Whether to loop idle animations
    var loopIdleAnimations: Bool = true

    // MARK: - Frame Queue Management

    /// Enqueue decoded frames for playback
    func enqueueFrames(_ frames: [DecodedFrame], forChunk chunkIndex: Int) {
        queueLock.lock()
        defer { queueLock.unlock() }

        frameQueue.append(contentsOf: frames)

        // Sort by sequence index
        frameQueue.sort { $0.sequenceIndex < $1.sequenceIndex }

        // Start playback if we have enough buffered
        if frameQueue.count >= bufferThreshold && !isPlaying {
            isPlaying = true
            mode = .talking
        }
    }

    /// Add a chunk to the queue
    func enqueueChunk(_ chunk: AnimationChunk) {
        chunkLock.lock()
        defer { chunkLock.unlock() }

        chunkQueue.append(chunk)
        chunkQueue.sort { $0.chunkIndex < $1.chunkIndex }
    }

    /// Mark a chunk as ready
    func markChunkReady(chunkIndex: Int) {
        chunkLock.lock()
        defer { chunkLock.unlock() }

        if let index = chunkQueue.firstIndex(where: { $0.chunkIndex == chunkIndex }) {
            chunkQueue[index].isReady = true
        }
    }

    /// Set the overlay position for lip sync
    func setOverlayPosition(_ position: CGPoint) {
        overlayPosition = position
    }

    /// Set the base frame
    func setBaseFrame(_ frame: UIImage?) {
        baseFrame = frame
    }

    /// Set base frame manager for idle animations
    func setBaseFrameManager(_ manager: BaseFrameManager?) {
        baseFrameManager = manager
    }

    /// Set animation mode
    func setMode(_ newMode: LegacyAnimationMode) {
        mode = newMode
    }

    /// Clear all queued frames
    func clearQueue() {
        queueLock.lock()
        frameQueue.removeAll()
        currentFrameIndex = 0
        queueLock.unlock()

        chunkLock.lock()
        chunkQueue.removeAll()
        chunkLock.unlock()

        isPlaying = false
    }

    /// Transition to idle
    func transitionToIdle() {
        mode = .idle
        // Keep current frames if any for smooth transition
    }

    // MARK: - Frame Retrieval

    /// Get the next frame for rendering (called by display link at 60fps)
    func getNextFrame() -> RenderFrame? {
        let currentTime = CACurrentMediaTime()

        // Check if enough time has passed for next frame based on mode
        let elapsed = currentTime - lastFrameTime
        guard elapsed >= mode.frameInterval else {
            // Return current frame without advancing
            return getCurrentRenderFrame(timestamp: currentTime)
        }

        lastFrameTime = currentTime

        // Handle idle mode with base frame manager
        if mode == .idle {
            return getIdleFrame(timestamp: currentTime)
        }

        queueLock.lock()
        defer { queueLock.unlock() }

        // Get current base frame from manager
        var currentBaseFrame = baseFrame
        if let manager = baseFrameManager {
            currentBaseFrame = manager.getCurrentIdleFrame()
        }

        // Get current overlay frame
        var overlayImage: UIImage? = nil

        if !frameQueue.isEmpty {
            if currentFrameIndex < frameQueue.count {
                overlayImage = frameQueue[currentFrameIndex].image
                currentFrameIndex += 1
            }

            // Handle end of queue
            if currentFrameIndex >= frameQueue.count {
                if mode == .talking {
                    // Talking done, check for more chunks or transition to idle
                    if hasMoreChunks() {
                        loadNextChunk()
                    } else {
                        // Stay on last frame briefly, then transition
                        currentFrameIndex = frameQueue.count - 1
                        overlayImage = frameQueue[currentFrameIndex].image
                    }
                }
            }
        }

        return RenderFrame(
            baseImage: currentBaseFrame,
            overlayImage: overlayImage,
            overlayPosition: overlayPosition,
            timestamp: currentTime
        )
    }

    /// Get frame for idle animation from base frame manager
    private func getIdleFrame(timestamp: CFTimeInterval) -> RenderFrame? {
        guard let manager = baseFrameManager else {
            // Fallback to static base frame
            return RenderFrame(
                baseImage: baseFrame,
                overlayImage: nil,
                overlayPosition: .zero,
                timestamp: timestamp
            )
        }

        // Advance and get next idle frame (10fps handled by mode.frameInterval)
        let idleFrame = manager.advanceFrame()

        return RenderFrame(
            baseImage: idleFrame ?? baseFrame,
            overlayImage: nil,
            overlayPosition: .zero,
            timestamp: timestamp
        )
    }

    /// Get current frame without advancing
    private func getCurrentRenderFrame(timestamp: CFTimeInterval) -> RenderFrame? {
        queueLock.lock()
        defer { queueLock.unlock() }

        var overlayImage: UIImage? = nil

        if !frameQueue.isEmpty && currentFrameIndex < frameQueue.count {
            overlayImage = frameQueue[currentFrameIndex].image
        } else if !frameQueue.isEmpty && currentFrameIndex > 0 {
            // Return last frame if we've gone past
            overlayImage = frameQueue[frameQueue.count - 1].image
        }

        return RenderFrame(
            baseImage: baseFrame,
            overlayImage: overlayImage,
            overlayPosition: overlayPosition,
            timestamp: timestamp
        )
    }

    // MARK: - Chunk Management

    private func hasMoreChunks() -> Bool {
        chunkLock.lock()
        defer { chunkLock.unlock() }
        return !chunkQueue.isEmpty && chunkQueue.first?.isReady == true
    }

    private func loadNextChunk() {
        chunkLock.lock()
        guard !chunkQueue.isEmpty, let chunk = chunkQueue.first, chunk.isReady else {
            chunkLock.unlock()
            return
        }

        let nextChunk = chunkQueue.removeFirst()
        chunkLock.unlock()

        queueLock.lock()
        // Keep some overlap for smooth transition
        let overlapFrames = min(5, frameQueue.count)
        if overlapFrames > 0 {
            frameQueue = Array(frameQueue.suffix(overlapFrames))
            currentFrameIndex = 0
        } else {
            frameQueue.removeAll()
            currentFrameIndex = 0
        }

        frameQueue.append(contentsOf: nextChunk.frames)
        overlayPosition = nextChunk.overlayPosition
        currentAnimationName = nextChunk.animationName
        queueLock.unlock()

        onChunkComplete?(nextChunk.chunkIndex)
    }

    // MARK: - Buffer Status

    /// Check if we have enough frames buffered to start playback
    var hasBufferedFrames: Bool {
        queueLock.lock()
        defer { queueLock.unlock() }
        return frameQueue.count >= minimumBufferFrames
    }

    /// Number of frames in queue
    var queuedFrameCount: Int {
        queueLock.lock()
        defer { queueLock.unlock() }
        return frameQueue.count
    }

    /// Number of chunks waiting
    var pendingChunkCount: Int {
        chunkLock.lock()
        defer { chunkLock.unlock() }
        return chunkQueue.count
    }

    /// Current playback progress (0.0 to 1.0)
    var playbackProgress: Float {
        queueLock.lock()
        defer { queueLock.unlock() }
        guard !frameQueue.isEmpty else { return 0 }
        return Float(currentFrameIndex) / Float(frameQueue.count)
    }

    // MARK: - Playback Control

    /// Start playback
    func play() {
        isPlaying = true
    }

    /// Pause playback
    func pause() {
        isPlaying = false
    }

    /// Reset to beginning
    func reset() {
        queueLock.lock()
        currentFrameIndex = 0
        queueLock.unlock()
        lastFrameTime = 0
    }

    /// Whether currently playing
    var isCurrentlyPlaying: Bool {
        return isPlaying
    }
}

// MARK: - Debug Helpers

extension AnimationEngine {
    /// Debug description of current state
    var debugDescription: String {
        return """
        AnimationEngine:
          Mode: \(mode)
          Playing: \(isPlaying)
          Frame: \(currentFrameIndex)/\(queuedFrameCount)
          Chunks pending: \(pendingChunkCount)
          Animation: \(currentAnimationName)
        """
    }
}
